File: stacks/basic_calculator.py
'''
LeetCode - problem 224
'''

    def calculate(self, s: str) -> int:

        res = 0
        num = 0
        operator = '+'
        stack = ['+'] 
        
        def execute_operator(stack, operator, num):
            current_status = stack[-1]
            if operator == current_status:   
                temp = num
            else:
                temp = -num
            return temp
        
        def calculate_current_substring_status(stack, operator):
            current_status = stack[-1]
            if operator == current_status: 
                temp = '+'
            else:
                temp = '-'
            return temp
        
        for char in s+"+":
            if char == ' ':
                continue
            elif char.isdigit():
                num = 10*num + int(char)
            elif char in "+-":
                res += execute_operator(stack, operator, num)
                operator = char
                num = 0
            elif char == "(":
                stack.append(calculate_current_substring_status(stack,operator))
                operator = '+'
            elif char == ")":
                res += execute_operator(stack, operator, num)
                num = 0
                stack.pop()
        return res

File: stacks/basic_calculator.cpp
//LeetCode - problem 224

    int helper(string s, int &i) {
        char pre_char = '+';
        long long now = 0;
        int ret = 0;

        for(; i<s.size(); i++) {
            if(s[i] == ' ')
                continue;
            if('0' <= s[i] && s[i] <= '9')
                now = now*10 + s[i] - '0';
            else {
                if(s[i] == '(')
                    now = helper(s, ++i);

                if(pre_char == '+')
                    ret += now;
                else if(pre_char == '-')
                    ret -= now;
                    
                if(i < s.size() && s[i] == ')') {
                    i++;
                    return ret;
                }
                
                pre_char = s[i];
                now = 0;
            }
        }

        return ret;
    }

    int calculate(string s) {
        s.push_back('#');
        int i = 0;
        return helper(s, i);
    }
